File: cpp/src/lexer/token.cpp
#include "txtcode/lexer/token.h"
#include <sstream>

namespace txtcode {

std::string Token::toString() const {
    std::ostringstream oss;
    
    switch (kind) {
        case TokenKind::Integer:
            oss << std::get<std::int64_t>(value.value);
            break;
        case TokenKind::Float:
            oss << std::get<double>(value.value);
            break;
        case TokenKind::String:
        case TokenKind::Identifier:
        case TokenKind::Comment:
        case TokenKind::MultiLineComment:
            oss << std::get<std::string>(value.value);
            break;
        case TokenKind::Boolean:
            oss << (std::get<bool>(value.value) ? "true" : "false");
            break;
        case TokenKind::Null:
            oss << "null";
            break;
        case TokenKind::Store:
            oss << "store";
            break;
        case TokenKind::Define:
            oss << "define";
            break;
        case TokenKind::Return:
            oss << "return";
            break;
        case TokenKind::If:
            oss << "if";
            break;
        case TokenKind::Else:
            oss << "else";
            break;
        case TokenKind::End:
            oss << "end";
            break;
        case TokenKind::Arrow:
            oss << "→";
            break;
        case TokenKind::Plus:
            oss << "+";
            break;
        case TokenKind::Minus:
            oss << "-";
            break;
        case TokenKind::Star:
            oss << "*";
            break;
        case TokenKind::Slash:
            oss << "/";
            break;
        case TokenKind::Equal:
            oss << "==";
            break;
        case TokenKind::NotEqual:
            oss << "!=";
            break;
        case TokenKind::Eof:
            oss << "EOF";
            break;
        default:
            oss << "<token>";
            break;
    }
    
    return oss.str();
}

} // namespace txtcode

File: src/lexer/token.rs
/// Token representation for the lexer
#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub value: String,
    pub span: (usize, usize),
}

impl Token {
    pub fn new(kind: TokenKind, value: String, span: (usize, usize)) -> Self {
        Self { kind, value, span }
    }
}

/// Token kinds for Txt-code
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    // Literals
    Integer,
    Float,
    String,
    InterpolatedString,
    Char,
    Boolean,
    Null,

    // Identifiers
    Identifier,
    Keyword,

    // Operators
    Arrow,         // -> or → (both supported)
    Plus,          // +
    Minus,         // -
    Star,          // *
    Slash,         // /
    Percent,       // %
    Power,         // **
    Assignment,    // = (single equals, for compound assignment or future use)
    Equal,         // ==
    NotEqual,      // !=
    Less,          // <
    Greater,       // >
    LessEqual,     // <=
    GreaterEqual,  // >=
    And,           // and
    Or,            // or
    Not,           // not
    NullCoalesce,  // ??
    OptionalChain, // ?.
    QuestionMark,  // ? (standalone)
    Pipe,          // |>
    Increment,     // ++
    Decrement,     // --

    // Compound assignment operators
    PlusAssign,    // +=
    MinusAssign,   // -=
    StarAssign,    // *=
    SlashAssign,   // /=
    PercentAssign, // %=
    PowerAssign,   // **=
    BitAndAssign,  // &=
    BitOrAssign,   // |=
    BitXorAssign,  // ^=

    // Bitwise
    BitAnd,     // &
    BitOr,      // |
    BitXor,     // ^
    LeftShift,  // <<
    RightShift, // >>
    BitNot,     // ~

    // Delimiters
    LeftParen,    // (
    RightParen,   // )
    LeftBrace,    // {
    RightBrace,   // }
    LeftBracket,  // [
    RightBracket, // ]
    Comma,        // ,
    Semicolon,    // ;
    Colon,        // :
    Dot,          // .
    Spread,       // ... (spread/rest operator)

    // Special
    Eof,
    Newline,
    Whitespace,
    Comment,
}
